File: Zuul_Java/Room.java
import java.util.Set;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;

/*
 * Class Room - a room in an adventure game.
 *
 * This class is the main class of the "World of Zuul" application. 
 * "World of Zuul" is a very simple, text based adventure game.  
 *
 * A "Room" represents one location in the scenery of the game.  It is 
 * connected to other rooms via exits.  For each existing exit, the room 
 * stores a reference to the neighboring room.
 * 
 * @author  Michael Kolling and David J. Barnes
 * @version 1.0 (February 2002)
 */

class Room 
{
    private String description;
    private HashMap exits;        // stores exits of this room.
    ArrayList<Item> items = new ArrayList<Item>();

    /**
     * Create a room described "description". Initially, it has no exits.
     * "description" is something like "in a kitchen" or "in an open court 
     * yard".
     */
    public Room(String description) 
    {
        this.description = description;
        exits = new HashMap();
    }

    /**
     * Define an exit from this room.
     */
    public void setExit(String direction, Room neighbor) 
    {
        exits.put(direction, neighbor);
    }

    /**
     * Return the description of the room (the one that was defined in the
     * constructor).
     */
    public String getShortDescription()
    {
        return description;
    }

    /**
     * Return a long description of this room, in the form:
     *     You are in the kitchen.
     *     Exits: north west
     */
    public String getLongDescription()
    {
        return "You are " + description + ".\n" + getExitString();
    }

    /**
     * Return a string describing the room's exits, for example
     * "Exits: north west".
     */
    private String getExitString()
    {
        String returnString = "Exits:";
        Set keys = exits.keySet();
        for(Iterator iter = keys.iterator(); iter.hasNext(); )
            returnString += " " + iter.next();
        returnString += "\nItems in the room:\n";
        returnString += getRoomItems();
        return returnString;
    }

    /**
     * Return the room that is reached if we go from this room in direction
     * "direction". If there is no room in that direction, return null.
     */
    public Room getExit(String direction) 
    {
        return (Room)exits.get(direction);
    }
    
    /*
     * Get items from the room
     */
    public Item getItem(int index) {
    	return items.get(index);
    }
    
    /*
     * Get item given item name
     * differentiate between two getItem methods based on parameter data type
     * arrays use .length without parentheses
     * arraylists use .size() with parentheses
     * strings use .length() with parentheses
     */
    public Item getItem(String itemName) {
    	// keyword null indicates that it's pointing to nothing in memory
    	// null pointer exception
    	for (int i = 0; i < items.size(); i++) {
			if (items.get(i).getDescription().equals(itemName)) {
				return items.get(i);
			}
		}
    	
    	return null;
    }
    
    /*
     * Remove items from the room
     */
    public void removeItem(String itemName) {
    	// keyword null indicates that it's pointing to nothing in memory
    	// null pointer exception
    	for (int i = 0; i < items.size(); i++) {
			if (items.get(i).getDescription().equals(itemName)) {
				items.remove(i); // remove item from items ArrayList
			}
		}
    }
    
    /*
     * Set a particular item in the room.
     */
    public void setItem(Item newItem) {
    	items.add(newItem);
    }
    
    /*
     * Get a description of the items in a room
     */
    public String getRoomItems() {
    	String output = "";
		for (int i = 0; i < items.size(); i++) { // must use .size() method to return length of ArrayList
			output += items.get(i).getDescription() + " ";
		}
		return output;
    }
} 

File: Room.cpp
#include <iostream>
#include <cstring>
#include "Room.h"
#include "Item.h"

using namespace std;

// Define constructor
Room::Room(char* description) {
  this->description = description;
}

// Define method to get room description
char* Room::getShortDescription() {
  return description;
}

// Define method to add exit from this room
void Room::setExit(char direction, Room* neighbor) {
  exits[direction] = neighbor;
}

/* Define method to print a description of where you are right now,
   direction of exits, and items in current room
 */
void Room::printLongDescription() {
  // room description
  cout << "You are " << getShortDescription() << "." << endl;

  // exits
  cout << "Exits:";
  for (auto it = exits.begin(); it != exits.end(); ++it) {
    cout << " " << it->first;
  }
  cout << endl;

  // items
  cout << "Items in the room:";
  for (auto it = items.begin(); it != items.end(); ++it) {
    cout << " " << (*it)->getDescription();
  }
  cout << endl;
}

// Define method to search for direction key in exits map
bool Room::searchDirection(char direction) {
  return (exits.find(direction) != exits.end());
}

// Define method to return destination room when moving in specified direction
Room* Room::getExitRoom(char direction) {
  return exits[direction];
}

// Define method to add item to room
void Room::setItem(Item* newItem) {
  items.push_back(newItem);
}


// Define method to return item given it's name
Item* Room::getItem(char* itemName) {
  for (int i = 0; i < items.size(); i++) {
    if (strcmp(items[i]->getDescription(), itemName) == 0) {
      return items[i];
    }
  }
  return nullptr;
}


// Define method to remove item from room given it's name
void Room::removeItem(char* itemName) {
  // use erase-remove idiom to mark items for removal
  auto removeItem = remove_if(items.begin(), items.end(), [&itemName](Item* ptr) {
    return (strcmp(itemName, ptr->getDescription()) == 0);
  });

  // erase marked items
  items.erase(removeItem, items.end());
}


// Define destructor
Room::~Room() {
  // clear items vector
  for (Item* ptr : items) {
    if (ptr != nullptr) { // validate ptr
      delete ptr;
    }
  }
  
  // do NOT delete description because it's a string literal
}
